src/pages/resultsView/ResultsViewPageStoreUtils.ts

```diff
@@ -410,23 +412,51 @@ export function doesQueryHaveCNSegmentData(
     const result : SampleAlteredMap = {};
     filteredAlterationData.forEach((element, key) => {
         //1: is not group
         if (element.mergedTrackOqlList === undefined) {
             const notGroupedOql = element.oql as OQLLineFilterOutput<AnnotatedExtendedAlteration>;                    
             const sampleKeys = _.map(notGroupedOql.data, (data) => data.uniqueSampleKey);
+            const unProfiledSampleKeys = samples.map((sample) => sample.uniqueSampleKey).filter((sampleKey) => {
+                // if not profiled in some genes molecular profile, then we think it is not profiled and will exclude this sample
+                return _.some(_.map(selectedMolecularProfileIds, (selectedMolecularProfileId) => {
+                    return isSampleProfiled(sampleKey, selectedMolecularProfileId, notGroupedOql.gene, coverageInformation);
+                }) , (profiled) => profiled === false);
+            });
             result[getSingleGeneResultKey(key, oqlQuery, notGroupedOql)] = samples.map((sample: Sample) => {
-                return sampleKeys.includes(sample.uniqueSampleKey);
+                if (unProfiledSampleKeys.includes(sample.uniqueSampleKey)) {
+                    return AlteredStatus.UNPROFILED;
+                } else if (sampleKeys.includes(sample.uniqueSampleKey)) {
+                    return AlteredStatus.ALTERED;
+                } else {
+                    return AlteredStatus.UNALTERED;
+                }
             });
         }
         //2: is group
         else {
             const groupedOql = element.oql as MergedTrackLineFilterOutput<AnnotatedExtendedAlteration>;
             const sampleKeys = _.map(_.flatten(_.map(groupedOql.list, (list) => list.data)), (data) => data.uniqueSampleKey);
+            const groupGenes = _.map(groupedOql.list, (oql) => oql.gene);
+            const unProfiledSampleKeys = samples.map((sample) => sample.uniqueSampleKey).filter((sampleKey) => {
+                // if not profiled in some genes molecular profile, then we think it is not profiled and will exclude this sample
+                return _.some(_.map(selectedMolecularProfileIds, (selectedMolecularProfileId) => {
+                    // if not profiled in every genes, then the sample is not profiled, or we think it is profiled
+                    return _.every(_.map(groupGenes, (gene) => {
+                        return isSampleProfiled(sampleKey, selectedMolecularProfileId, gene, coverageInformation);
+                    }), (profiled) => profiled === false);
+                }) , (notProfiled) => notProfiled === true);
+            });
             result[getMultipleGeneResultKey(groupedOql)] = samples.map((sample: Sample) => {
-                return sampleKeys.includes(sample.uniqueSampleKey);
+                if (unProfiledSampleKeys.includes(sample.uniqueSampleKey)) {
```

Comment:
i just realized we should optimize this.    you should first make a map from the array . {  uniqueSampleId: true }
then in loop you can do simple lookup instead of using expensive array.includes.  this will probably be 20 times faster.   since we can tens of thousands of samples, this is actually important.

Reply:
i would add that you should only do this when you know N is going to be big and operation itself is expensive.  this is an obvious case.   you do the work of indexing the samples once up front and that saves you from using .includes  over and over again. (includes is just doing a full loop)

Reply:
Thanks! I think I can make `sampleKeys` and `unProfiledSampleKeys` to `sampleKeysMap` and `unProfiledSampleKeysMap`

Reply:
a handy method for this is `_.keyBy(listOfStrings)` returns `{ string1:"string1", string2:"string2", ...}` then you can check `sample in myMap`